Update the developer guide to add a section to run remote backed storage from dev environment (#16223)
opensearch-project/OpenSearch: DEVELOPER_GUIDE.md

git secrets --register-aws
```

#### Run remote backed storage from dev environment

To run the remote backed storage from the dev/local environment - Please do add the below to run.gradle file.

```
testClusters {
   runTask {
    // Add following lines to enable remote store
    setting 'node.attr.remote_store.segment.repository', '<repository_name>'
    setting 'node.attr.remote_store.translog.repository', '<repository_name>'
    setting 'node.attr.remote_store.state.repository', '<repository_name>'
    setting 'node.attr.remote_store.repository.my-repository.type', '<type>'
    setting 'node.attr.remote_store.repository.my-repository.settings.bucket', '<bucket_name>'
    setting 'node.attr.remote_store.repository.my-repository.settings.base_path', '<base_path>'
    setting 'node.attr.remote_store.repository.my-repository.settings.region', '<region>'

    // Add AWS credentials to the keystore

>     keystore 's3.client.default.access_key', '<access_key>'

Reviewer: I don't think storing credentials in this file should be a recommended practice. 

How about reading from env variables instead as an alternative unless there are better ways:

```
keystore 's3.client.default.access_key', System.getenv('AWS_ACCESS_KEY_ID')
keystore 's3.client.default.secret_key', System.getenv('AWS_SECRET_ACCESS_KEY')
keystore 's3.client.default.session_token', System.getenv('AWS_SESSION_TOKEN')
```

```
export AWS_ACCESS_KEY_ID=<access_key>
export AWS_SECRET_ACCESS_KEY=<secret_key>
export AWS_SESSION_TOKEN=<session_token>

```

In that way onlly authorized users are able to run this script.

Author: I understand that storing credentials directly isn't the recommended approach, and I agree. You can definitely export these values and access them through system variables. My goal here is to demonstrate how to run OpenSearch locally while connecting to remote storage. 

The reason I’m using them directly in the file is that my tokens [access_key, secret, token] are temporarily generated and have a validity of only 45 minutes. 

Made the recommended changes.

git secrets --register-aws
```

> #### Remote Backed Storage

Reviewer 2: The developer guide has a TOC on the top that needs to be re-generated/updated.

![Screenshot 2024-10-16 at 6 40 13 AM](https://github.com/user-attachments/assets/8b712d0e-3321-43b6-b04d-87a37aa9c7c0)

Add the following settings to [gradle/run.gradle](gradle/run.gradle) to enable remote-backed S3 storage in the dev/local environment.

```
testClusters {
   runTask {

>     // Add following lines to enable remote store

Reviewer 2: To be consistent with the text below.

```suggestion
    // Enable remote store
```

    setting 'node.attr.remote_store.segment.repository', '<repository_name>'
    setting 'node.attr.remote_store.translog.repository', '<repository_name>'
    setting 'node.attr.remote_store.state.repository', '<repository_name>'
    setting 'node.attr.remote_store.repository.my-repository.type', '<type>'
    setting 'node.attr.remote_store.repository.my-repository.settings.bucket', '<bucket_name>'
    setting 'node.attr.remote_store.repository.my-repository.settings.base_path', '<base_path>'
    setting 'node.attr.remote_store.repository.my-repository.settings.region', '<region>' # e.g. us-west-2

> 

Reviewer 2: Nit: remove one of the empty lines.

    // Add AWS credentials to the keystore
    keystore 's3.client.default.access_key', System.getenv('AWS_ACCESS_KEY_ID')
    keystore 's3.client.default.secret_key', System.getenv('AWS_SECRET_ACCESS_KEY')
    keystore 's3.client.default.session_token', System.getenv('AWS_SESSION_TOKEN')
```

> Then run by giving the required plugin as parameter

Reviewer 2: ```suggestion
Then run by specifying the required plugin as parameter.
```

```

> Exporting values to use as system variable. 

Reviewer 2: Remove this line, it's text and can't be copy-pasted.

export AWS_ACCESS_KEY_ID=<access_key>
export AWS_SECRET_ACCESS_KEY=<secret_key>
export AWS_SESSION_TOKEN=<session_token>

> Example: plugin can be 'repository-s3'

Reviewer 2: In this section it _has_ to be `repository-s3`, just remove this line.

git secrets --register-aws
```

> #### Run remote backed storage from dev environment

Reviewer 2: Make it consistent with other sections.

```suggestion
#### Remote Backed Storage
```

Since all the examples below are S3, I'd add a subsection for that, giving an opportunity add others later.

```
##### S3
```

Also update TOC on the top of the doc, please.

> To run the remote backed storage from the dev/local environment - Please do add the below to run.gradle file.

Reviewer 2: ```suggestion
Add the following settings to [gradle/run.gradle](gradle/run.gradle) to enable remote-backed S3 storage in the dev/local environment.
```

```
testClusters {
   runTask {
    // Add following lines to enable remote store
    setting 'node.attr.remote_store.segment.repository', '<repository_name>'
    setting 'node.attr.remote_store.translog.repository', '<repository_name>'
    setting 'node.attr.remote_store.state.repository', '<repository_name>'
    setting 'node.attr.remote_store.repository.my-repository.type', '<type>'
    setting 'node.attr.remote_store.repository.my-repository.settings.bucket', '<bucket_name>'
    setting 'node.attr.remote_store.repository.my-repository.settings.base_path', '<base_path>'

>     setting 'node.attr.remote_store.repository.my-repository.settings.region', '<region>'

Reviewer 2: Maybe add examples?

```suggestion
    setting 'node.attr.remote_store.repository.my-repository.settings.region', '<region>' # e.g. us-west-2
```

    // Add AWS credentials to the keystore
    keystore 's3.client.default.access_key', System.getenv('AWS_ACCESS_KEY_ID')
    keystore 's3.client.default.secret_key', System.getenv('AWS_SECRET_ACCESS_KEY')

>     keystore 's3.client.default.session_token', System.getenv('AWS_SESSION_TOKEN')

Reviewer 3: Should these be wrapped in a null check?  What happens if the env is not set?

Reviewer 2: `System.getenv` returns `null` which I think is fine for a dev environment. When you try to exercise the API you'll get access denied errors as expected.

>      testDistribution = 'archive'

Reviewer 2: Either line this text up to the rest of the text or just remove this part since it's already there and is not relevant to our example.

Author: Removed - as suggested.

     if (numZones > 1) numberOfZones = numZones
     if (numNodes > 1) numberOfNodes = numNodes
```
Then run by giving the required plugin as parameter
```
Exporting values to use as system variable. 
export AWS_ACCESS_KEY_ID=<access_key>
export AWS_SECRET_ACCESS_KEY=<secret_key>
export AWS_SESSION_TOKEN=<session_token>

Example: plugin can be 'repository-s3'

> ./gradlew run -PinstalledPlugins="['<plugin>']" --debug-jvm

Reviewer 2: ```suggestion
./gradlew run -PinstalledPlugins="['repository-s3']"
```

(` --debug-jvm` is not relevant for this part)
